x-pack/plugins/security_solution/server/lib/detection_engine/signals/build_bulk_body.ts

```diff
@@ -164,3 +166,54 @@ export const buildSignalFromEvent = (
   };
   return signalHit;
 };
+
+export const objectArrayIntersection = (objects: object[]) => {
+  if (objects.length === 0) {
+    return undefined;
+  } else if (objects.length === 1) {
+    return objects[0];
+  } else {
+    return objects
+      .slice(1)
+      .reduce(
+        (acc: object | undefined, obj): object | undefined => objectPairIntersection(acc, obj),
+        objects[0]
+      );
+  }
+};
+
+export const objectPairIntersection = (a: object | undefined, b: object | undefined) => {
```

Comment:
This function has the potential for a lot of tricky edge cases with the `typeof` and `instanceof` checks so it has a number of associated unit tests. If anyone has suggestions on how to do this in a cleaner way I'd be happy to discuss.